Tutorial: access review of assignments to privileged roles (#14283)
microsoftgraph/microsoft-graph-docs: concepts/tutorial-accessreviews-roleassignments.md

      :::image type="content" source="/graph/images/GE-Permissions/User.ReadWrite.All-consent.png" alt-text="Consent to Microsoft Graph permissions." border="true":::

>[!NOTE]
>The response objects shown in this tutorial might be shortened for readability.

## Step 1: Create an access review of privileged role assignments

In this tutorial, we create a recurring access review of both *active* and *eligible* assignments to the User Administrator role. An **accessReviewScheduleDefinition** can be used to define the access review of multiple principle types (users and groups, or service principals) to only one privileged role. To review access to multiple privileged roles, create separate accessReviewScheduleDefinition objects.

The following access review schedule definition has the following settings:

+ The principals (**principalScopes**) whose access will be reviewed are groups and users, and the resource (**resourceScopes**) is the User Administrator role.
+ The scope (**resourceScopes**) of the review is on both active and eligible User Administrator assignments.
+ You as the user in an Identity Governance Administrator role will be the reviewer.

> + The user in the Global Administrator role is the fallback reviewer. If the reviewers haven't recorded decisions when the review instance nears expiry, the fallback reviewer is notified of the pending access review for action.

Author: Well noted. Applied the change. Though I'll ping you offline to understand this better.

+ The approver must provide justification before they approve access to the privileged role.
+ The default decision is `Deny` when the main and fallback reviewers don't respond to the access review request before the instance expires.
+ **autoApplyDecisionsEnabled** isn't set and defaults to `false`. In this case, after the review completes, the decisions aren't automatically applied but you must manually apply them.
+ The review recurs every three months over a period of three days and doesn't end.

### Request

In the following request, replace the following values:

+ `4562bcc8-c436-4f95-b7c0-4f8ce89dca5e` with the value of your user identifier.

> + `77f8b7b6-1119-4a86-b4c3-e7d68e990424` with the value of the global administrator's user identifier.

Author: On it locally.

<!-- {
  "blockType": "request",
  "name": "tutorial-accessreviewsroleassignments-create"
}-->
```http
POST https://graph.microsoft.com/v1.0/identityGovernance/accessReviews/definitions
Content-type: application/json

{
    "displayName": "Review access of users and groups to privileged roles",
    "descriptionForAdmins": "Review access of users and groups to privileged roles",
    "scope": {
        "@odata.type": "#microsoft.graph.principalResourceMembershipsScope",
        "principalScopes": [
            {
                "@odata.type": "#microsoft.graph.accessReviewQueryScope",
                "query": "/users",
                "queryType": "MicrosoftGraph"
            },
            {
                "@odata.type": "#microsoft.graph.accessReviewQueryScope",
                "query": "/groups",
                "queryType": "MicrosoftGraph"
            }

>         ],

Reviewer: bump @shubhamguptacal !

            "decision": "NotReviewed",
            "justification": "",
            "appliedDateTime": null,
            "applyResult": "New",
            "recommendation": "NoInfoAvailable",
            "principalLink": "https://graph.microsoft.com/v1.0/users/99e44b05-c10b-4e95-a523-e2732bbaba1e",
            "resourceLink": null,
            "resource": null,
            "reviewedBy": {
                "id": "00000000-0000-0000-0000-000000000000",
                "displayName": "",
                "type": null,
                "userPrincipalName": ""
            },
            "appliedBy": {
                "id": "00000000-0000-0000-0000-000000000000",
                "displayName": "",
                "type": null,
                "userPrincipalName": ""
            },
            "principal": {
                "@odata.type": "#microsoft.graph.userIdentity",
                "id": "99e44b05-c10b-4e95-a523-e2732bbaba1e",
                "displayName": "Adele Vance",
                "type": "user",
                "userPrincipalName": "AdeleV@M365x010717.OnMicrosoft.com",
                "lastUserSignInDateTime": ""
            }
        }
    ]
}
```

> You can now record or submit your decisions for the access review instances.

Reviewer: ```suggestion
As the reviewer, you can now submit your decisions for the access review instance.
```

    "justification": "Adele Vance should join an allowed group to have continued access to the User Administrator role. Refer to the company policy '#132487: Privileged roles' for more details."
}
```

### Response

```
HTTP/1.1 204 No Content
```

When you retrieve the access review decisions, they have the following settings:
+ The access review decision for the IT Helpdesk group is `Approve` while for Adele is `Deny`.
+ The reviewedBy object contains your details as the reviewer.
+ The applyResult is `New` meaning the decisions haven't been applied.

Even though you've recorded all the decisions that were pending for this accessReviewInstance, these decisions haven't been applied to the resource and principal objects. For example, Adele still has access to the privileged role. This is because the **autoApplyDecisionsEnabled** was set to `false`, you haven't stopped the review, nor has the accessReviewInstance period expired and autocompleted.

In this tutorial, you won't stop the accessReviewInstance manually but, you'll let the accessReviewInstance expire and autocomplete and then apply the access review decisions.

## Step 5: Apply access review decisions

> As an admin, after the **status** of the accessReviewInstance is set to `Completed`, you can apply the access review decisions.

Reviewer: ```suggestion
As an admin, after the 3 day duration has passed and the **status** of the accessReviewInstance is set to `Completed`, you can apply the access review decisions.
```

```http
POST https://graph.microsoft.com/v1.0/identityGovernance/accessReviews/definitions/a13c348b-dc1c-47b5-8c58-b0d12b35a18b/instances/f282e54c-4eed-47a4-be9a-da00ab38aa8f/applyDecisions
```

### Response

```
HTTP/1.1 204 No Content
```

Adele has now lost access to the User Administrator role while the IT Support group has continued access.

## Step 6: Retrieve access review decisions

Contoso's auditors are reviewing of all decisions to grant or deny access to privileged roles in the organization. You'll retrieve access review decision logs for all access reviews scoped to privileged roles. In this example, you'll retrieve decision logs for the accessReviewScheduleDefinition you created in Step 1.

### Request

```http
GET https://graph.microsoft.com/v1.0/identityGovernance/accessReviews/definitions/a13c348b-dc1c-47b5-8c58-b0d12b35a18b/instances/f282e54c-4eed-47a4-be9a-da00ab38aa8f/decisions
```

### Response

The following response object is different from the response object you received in Step 3 with the following settings:
+ The access review decision for IT Helpdesk is Approve while for Adele is Deny
+ The reviewedBy object contains your details as the reviewer

> + The applyResult is New meaning the decisions haven't yet been applied. You must manually apply the decisions.

Reviewer: But didn't we just apply the decisions in the step above?

Author: Q: Good catch. Hadn't run the query for applying decisions for Adele.

            "appliedDateTime": null,
            "applyResult": "New",
            "recommendation": "NoInfoAvailable",
            "principalLink": "https://graph.microsoft.com/v1.0/users/99e44b05-c10b-4e95-a523-e2732bbaba1e",
            "resourceLink": null,
            "resource": null,
            "reviewedBy": {
                "id": "4562bcc8-c436-4f95-b7c0-4f8ce89dca5e",
                "displayName": "MOD Administrator",
                "type": null,
                "userPrincipalName": "admin@M365x010717.onmicrosoft.com"
            },
            "appliedBy": {
                "id": "00000000-0000-0000-0000-000000000000",
                "displayName": "",
                "type": null,
                "userPrincipalName": ""
            },
            "principal": {
                "@odata.type": "#microsoft.graph.userIdentity",
                "id": "99e44b05-c10b-4e95-a523-e2732bbaba1e",
                "displayName": "Adele Vance",
                "type": "user",
                "userPrincipalName": "AdeleV@M365x010717.OnMicrosoft.com",
                "lastUserSignInDateTime": ""
            }
        }
    ]
}
```

> <!---

Reviewer: i think we should definitely include this part

Author: I think it best that we use the recent changes thanks to Levi's PR? Let me retest.

## Step 6: Retrieve access review history definitions

> Contoso's auditors also want to review the access review history for the last quarter. In this example, you'll generate an access review history report for all acecssReviewScheduleDefinitions scoped to directory role assignments (roleAssignmentScheduleInstances). In this query, the **decisions** property is empty and therefore defaults to include all decisions in the history report.

Reviewer: ```suggestion
Contoso's auditors also want to review the access review history for the last quarter. In this example, you'll generate an access review history report for all accessReviewScheduleDefinitions scoped to directory role assignments (roleAssignmentScheduleInstances). In this query, the **decisions** property is empty and therefore defaults to include all decisions in the history report.
```

---
title: "Tutorial: Use the access reviews API to review access to privileged roles"
description: "Use the access reviews API to review access to privileged roles"
author: "FaithOmbongi"
ms.localizationpriority: medium
ms.prod: "governance"
---

# Tutorial: Use the access reviews API to review access to privileged roles

Contoso Limited is a growing service provider that has delegated various Azure AD administrator privileges to users, groups, and service principals in the organization. The company needs to ensure only the right assignees have access to privileged roles. The system auditors should also audit the access review history to report on the effectiveness of Contoso's internal controls.

In this tutorial, you'll use the access reviews API to create a recurring access review of users and groups with access to privileged roles. This access includes both active and eligible roles.

## Prerequisites

To complete this tutorial, you need the following resources and privileges:

+ A working Azure AD tenant with an Azure AD Premium P2 or EMS E5 license enabled.

> + Sign in to [Graph Explorer](https://developer.microsoft.com/graph/graph-explorer) as a user in an Identity Governance Administrator role.

Reviewer: ```suggestion
+ Sign in to [Graph Explorer](https://developer.microsoft.com/graph/graph-explorer) as a user in a Privileged Role Administrator role.
```

> + Principals assigned to a privileged role that is the scope of your access review. To assign privileged roles, see [Tutorial: Use the Privileged Identity Management (PIM) API to assign Azure AD roles](/graph/tutorial-assign-azureadroles).

Reviewer: ```suggestion
+ Principals assigned to a privileged role. These assignments will be the scope of your access review. To assign privileged roles, see [Tutorial: Use the Privileged Identity Management (PIM) API to assign Azure AD roles](/graph/tutorial-assign-azureadroles).
```

### Request

```http
POST https://graph.microsoft.com/v1.0/identityGovernance/accessReviews/definitions/a13c348b-dc1c-47b5-8c58-b0d12b35a18b/instances/f282e54c-4eed-47a4-be9a-da00ab38aa8f/applyDecisions
```

### Response

```
HTTP/1.1 204 No Content
```

Adele has now lost access to the User Administrator role while the IT Support group has continued access.

## Step 6: Retrieve access review decisions

Contoso's auditors are reviewing of all decisions to grant or deny access to privileged roles in the organization. You'll retrieve access review decision logs for all access reviews scoped to privileged roles. In this example, you'll retrieve decision logs for the **accessReviewScheduleDefinition** you created in Step 1.

### Request

```http
GET https://graph.microsoft.com/v1.0/identityGovernance/accessReviews/definitions/a13c348b-dc1c-47b5-8c58-b0d12b35a18b/instances/f282e54c-4eed-47a4-be9a-da00ab38aa8f/decisions
```

### Response

The following response object is different from the response object you received in Step 3 with the following settings:
+ The access review decision for IT Helpdesk is Approve while for Adele is Deny

> + The reviewedBy object contains your details as the reviewer

Reviewer: Also add that the deny result on Adele as applied, so the appliedBy property is filled on her decision
